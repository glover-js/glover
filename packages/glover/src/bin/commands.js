// @flow

export const brew = {
  cmd: ['brew [recipe]', '$0'],
  desc: 'Brew a recipe',
  builder: (yargs: any) => {
    yargs.positional('recipe', {
      type: 'string',
      describe: 'The recipe you want to brew',
    });
    yargs.option('skipInstall', {
      type: 'boolean',
      describe: 'Skip npm/yarn install',
      default: false,
    });
    yargs.option('skipIngredient', {
      type: 'array',
      describe: 'Skip ingredients in the recipe',
      default: [],
    });
    yargs.option('skipStep', {
      type: 'array',
      describe: 'Skip steps in the recipe',
      default: [],
    });
  },
};

export const pick = {
  cmd: 'pick [ingredients...]',
  desc: 'Pick ingredients from a recipe',
  builder: (yargs: any) => {
    yargs.positional('ingredients', {
      type: 'array',
      describe: 'The ingredients you want to pick',
    });
    yargs.option('prompt', {
      alias: 'p',
      type: 'boolean',
      describe: '',
      default: false,
    });
    yargs.option('output', {
      alias: 'o',
      type: 'string',
      describe: 'Output path',
      default: process.cwd(),
    });
  },
};

export const compose = {
  cmd: 'compose [name] [ingredients...]',
  desc: 'compose a new recipe',
  builder: (yargs: any) => {
    yargs.positional('name', {
      type: 'string',
      describe: 'The name of your recipe',
    });
    yargs.positional('ingredients', {
      type: 'array',
      describe: 'Ingredients you want to include in your recipe',
    });
    yargs.option('extend', {
      alias: 'e',
      type: 'array',
      describe: 'Recipe(s) you want to extend',
      default: [],
    });
    yargs.option('output', {
      alias: 'o',
      type: 'string',
      describe: 'Output path',
      default: process.cwd(),
    });
  },
};
