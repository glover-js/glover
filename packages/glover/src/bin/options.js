// @flow
export const dryRun = {
  key: 'dryRun',
  options: {
    type: 'boolean',
    describe: 'Run the command without generating any files',
    default: false,
  },
};

export const silent = {
  key: 'silent',
  options: {
    type: 'boolean',
    describe: 'Prevent the command from printing anything to the console',
    default: false,
  },
};

export const noPrompt = {
  key: 'noPrompt',
  options: {
    type: 'boolean',
    describe: 'Prevent the command from prompting for values',
    default: false,
  },
};

export const debug = {
  key: 'debug',
  options: {
    type: 'boolean',
    describe: 'Print debugging info',
    default: false,
  },
};
