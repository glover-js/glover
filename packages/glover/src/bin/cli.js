// @flow
import cli from 'yargs';
import { brew, pick, compose } from './commands';
import { dryRun, silent, noPrompt, debug } from './options';
import run from '..';

export function parseArgv(argv: Array<string>): Promise<any> {
  return new Promise((resolve: any, reject: any) => {
    cli
      .scriptName('glover')
      .command(brew.cmd, brew.desc, brew.builder)
      .command(pick.cmd, pick.desc, pick.builder)
      .command(compose.cmd, compose.desc, compose.builder)
      .option(dryRun.key, dryRun.options)
      .option(silent.key, silent.options)
      .option(noPrompt.key, noPrompt.options)
      .option(debug.key, debug.options)
      .wrap(120)
      .parse(argv.slice(2), (err: any, parsedArgs: any, output: any) => {
        if (err) {
          return reject(err);
        }
        return resolve({
          arguments: parsedArgs,
          command: parsedArgs._[0] || 'brew',
          output,
          argv,
        });
      });
  });
}

if (process.env.NODE_ENV !== 'test') {
  parseArgv(process.argv)
    .then(args => {
      if (args.output) {
        console.log(args.output); // eslint-disable-line no-console
        process.exit(0);
      }
      return run(args);
    })
    .catch(err => {
      console.log(err); // eslint-disable-line no-console
      process.exit(1);
    });
}
