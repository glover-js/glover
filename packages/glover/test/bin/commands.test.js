import { brew, pick, compose } from '../../src/bin/commands';

describe('Tests for commands', () => {
  describe('brew', () => {
    it('exposes the correct cmd and desc properties', () => {
      expect(brew.cmd).toEqual(['brew [recipe]', '$0']);
      expect(brew.desc).toEqual('Brew a recipe');
    });
    it('exposes a correct builder function', () => {
      const yargs = { positional: jest.fn(), option: jest.fn() };
      brew.builder(yargs);
      expect(yargs.positional.mock.calls).toMatchSnapshot();
      expect(yargs.option.mock.calls).toMatchSnapshot();
    });
  });
  describe('pick', () => {
    it('exposes the correct cmd and desc properties', () => {
      expect(pick.cmd).toEqual('pick [ingredients...]');
      expect(pick.desc).toEqual('Pick ingredients from a recipe');
    });
    it('exposes a correct builder function', () => {
      const yargs = { positional: jest.fn(), option: jest.fn() };
      pick.builder(yargs);
      expect(yargs.positional.mock.calls).toMatchSnapshot();
      expect(yargs.option.mock.calls).toMatchSnapshot();
    });
  });
  describe('compose', () => {
    it('exposes the correct cmd and desc properties', () => {
      expect(compose.cmd).toEqual('compose [name] [ingredients...]');
      expect(compose.desc).toEqual('compose a new recipe');
    });
    it('exposes a correct builder function', () => {
      const yargs = { positional: jest.fn(), option: jest.fn() };
      compose.builder(yargs);
      expect(yargs.positional.mock.calls).toMatchSnapshot();
      expect(yargs.option.mock.calls).toMatchSnapshot();
    });
  });
});
