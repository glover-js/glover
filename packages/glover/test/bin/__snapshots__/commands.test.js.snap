// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for commands brew exposes a correct builder function 1`] = `
Array [
  Array [
    "recipe",
    Object {
      "describe": "The recipe you want to brew",
      "type": "string",
    },
  ],
]
`;

exports[`Tests for commands brew exposes a correct builder function 2`] = `
Array [
  Array [
    "skipInstall",
    Object {
      "default": false,
      "describe": "Skip npm/yarn install",
      "type": "boolean",
    },
  ],
  Array [
    "skipIngredient",
    Object {
      "default": Array [],
      "describe": "Skip ingredients in the recipe",
      "type": "array",
    },
  ],
  Array [
    "skipStep",
    Object {
      "default": Array [],
      "describe": "Skip steps in the recipe",
      "type": "array",
    },
  ],
]
`;

exports[`Tests for commands compose exposes a correct builder function 1`] = `
Array [
  Array [
    "name",
    Object {
      "describe": "The name of your recipe",
      "type": "string",
    },
  ],
  Array [
    "ingredients",
    Object {
      "describe": "Ingredients you want to include in your recipe",
      "type": "array",
    },
  ],
]
`;

exports[`Tests for commands compose exposes a correct builder function 2`] = `
Array [
  Array [
    "extend",
    Object {
      "alias": "e",
      "default": Array [],
      "describe": "Recipe(s) you want to extend",
      "type": "array",
    },
  ],
  Array [
    "output",
    Object {
      "alias": "o",
      "default": "/Users/ollelauribostrom/Desktop/glover",
      "describe": "Output path",
      "type": "string",
    },
  ],
]
`;

exports[`Tests for commands pick exposes a correct builder function 1`] = `
Array [
  Array [
    "ingredients",
    Object {
      "describe": "The ingredients you want to pick",
      "type": "array",
    },
  ],
]
`;

exports[`Tests for commands pick exposes a correct builder function 2`] = `
Array [
  Array [
    "prompt",
    Object {
      "alias": "p",
      "default": false,
      "describe": "",
      "type": "boolean",
    },
  ],
  Array [
    "output",
    Object {
      "alias": "o",
      "default": "/Users/ollelauribostrom/Desktop/glover",
      "describe": "Output path",
      "type": "string",
    },
  ],
]
`;
